File: engine/validate_cli.py
"""Validate job definitions without executing anything.

Used by .github/workflows/validate.yml on push/PR to jobs/**, and safe to run
locally any time you add or edit a job.
"""
from __future__ import annotations

import sys
from pathlib import Path

from . import discovery, schema
from .executors.script_executor import ScriptSecurityError, resolve_interpreter, resolve_script_path

REPO_ROOT = Path(__file__).resolve().parent.parent
JOBS_ROOT = REPO_ROOT / "jobs"


def main() -> int:
    try:
        jobs = discovery.discover_jobs(JOBS_ROOT)
    except (discovery.DiscoveryError, schema.ValidationError) as e:
        print("Job validation failed:\n", file=sys.stderr)
        print(str(e), file=sys.stderr)
        return 1

    # Schema validation can't check the filesystem — catch unsafe/missing
    # script paths here so a bad job fails on PR, not at 3am in production.
    script_errors = []
    for job in jobs:
        if job.type != "script":
            continue
        try:
            resolve_script_path(REPO_ROOT, job.script.path)
            resolve_interpreter(job.script)
        except ScriptSecurityError as e:
            script_errors.append(f"{job.file_path}: {e}")
    if script_errors:
        print("Job validation failed:\n", file=sys.stderr)
        print("\n".join(script_errors), file=sys.stderr)
        return 1

    print(f"OK: {len(jobs)} job(s) validated under {JOBS_ROOT}")
    for job in sorted(jobs, key=lambda j: j.id):
        status = "enabled" if job.enabled else "disabled"
        print(f"  - {job.id} [{job.type}, {status}] schedule='{job.schedule}' tz={job.timezone}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/engine/validate-cli.js
"use strict";

const path = require("path");
const discovery = require("./discovery");
const schema = require("./schema");
const {
  ScriptSecurityError,
  resolveInterpreter,
  resolveScriptPath,
} = require("./executors/script-executor");

const REPO_ROOT = path.resolve(__dirname, "../..");
const JOBS_ROOT = path.join(REPO_ROOT, "jobs");

function main() {
  let jobs;
  try {
    jobs = discovery.discoverJobs(JOBS_ROOT);
  } catch (err) {
    if (err instanceof discovery.DiscoveryError || err instanceof schema.ValidationError) {
      console.error("Job validation failed:\n");
      console.error(err.message);
      return 1;
    }
    throw err;
  }

  const scriptErrors = [];
  for (const job of jobs) {
    if (job.type !== "script") continue;
    try {
      resolveScriptPath(REPO_ROOT, job.script.path);
      resolveInterpreter(job.script);
    } catch (err) {
      if (err instanceof ScriptSecurityError) {
        scriptErrors.push(`${job.file_path}: ${err.message}`);
      } else {
        throw err;
      }
    }
  }
  if (scriptErrors.length) {
    console.error("Job validation failed:\n");
    console.error(scriptErrors.join("\n"));
    return 1;
  }

  console.log(`OK: ${jobs.length} job(s) validated under ${JOBS_ROOT}`);
  for (const job of [...jobs].sort((a, b) => a.id.localeCompare(b.id))) {
    const status = job.enabled ? "enabled" : "disabled";
    console.log(`  - ${job.id} [${job.type}, ${status}] schedule='${job.schedule}' tz=${job.timezone}`);
  }
  return 0;
}

if (require.main === module) {
  process.exitCode = main();
}

module.exports = { main };
